File: examples/server/02_credentials/01_issuing_credentials/05_comprehensive/go/main.go
// Package main demonstrates comprehensive credential issuance capabilities using the Self SDK.
//
// This example shows how to:
// - Initialize Self accounts for issuer and holder roles using the core SDK
// - Create various types of verifiable credentials using the underlying credential module
// - Attach evidence/files to credentials for enhanced verification
// - Handle complex nested claims and data structures
// - Create verifiable presentations from credentials
// - Manage asset uploads and downloads for evidence using the core SDK
//
// The Self SDK provides decentralized identity and verifiable credential capabilities,
// allowing entities to issue, hold, and verify credentials without requiring
// centralized authorities while maintaining cryptographic integrity and privacy.
//
// 🎯 CREDENTIAL CAPABILITIES DEMONSTRATED:
// • Basic credential creation (Email verification)
// • Multi-claim credentials (Profile information)
// • Custom credentials with file evidence (Certifications)
// • Complex nested data structures (Organization credentials)
// • Credential builder pattern usage
// • Asset/evidence management (file uploads)
// • Verifiable presentation creation
// • Request/response handling workflows
//
// 🔧 KEY SDK COMPONENTS SHOWCASED:
// • account.New() - Account initialization and configuration using core SDK
// • credential.NewCredential() - Direct credential construction using core API
// • object.New() - Evidence and file attachment management using core SDK
// • credential.NewPresentation() - Verifiable presentation creation using core API
// • account.CredentialIssue() - Direct credential issuance using core SDK
//
// 📚 EDUCATIONAL PROGRESSION:
// The examples progress from simple to complex, building understanding:
// 1. Basic Email Credential - Simplest form with minimal claims
// 2. Profile Credential - Multiple claims in a single credential
// 3. Custom Credential with Evidence - File attachments and presentations
// 4. Organization Credential - Complex nested data and arrays
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"fmt"
	"log"
	"time"

	"github.com/joinself/self-go-sdk/account"
	"github.com/joinself/self-go-sdk/credential"
	"github.com/joinself/self-go-sdk/object"
)

const (
	// Configuration constants for demo setup
	issuerStorageDir = "./issuer_storage"
	holderStorageDir = "./holder_storage"
)

func main() {
	fmt.Println("Comprehensive Credential Issuance Demo")
	fmt.Println("======================================")

	// Initialize issuer and holder clients
	issuerAccount, holderAccount := setupClients()
	defer issuerAccount.Close()
	defer holderAccount.Close()

	// Display account information
	displayAccountInfo(issuerAccount, holderAccount)

	// Credential issuance examples
	fmt.Println("CREDENTIAL ISSUANCE EXAMPLES")
	fmt.Println("============================")

	demonstrateBasicCredential(issuerAccount, holderAccount)
	demonstrateProfileCredential(issuerAccount, holderAccount)
	demonstrateCustomCredentialWithEvidence(issuerAccount, holderAccount)
	demonstrateOrganizationCredential(issuerAccount, holderAccount)

	fmt.Println("✅ All examples completed successfully!")
}

// generateStorageKey creates a cryptographically secure 32-byte key
func generateStorageKey(seed string) []byte {
	key := make([]byte, 32)
	if _, err := rand.Read(key); err != nil {
		// Fallback to deterministic key generation if crypto/rand fails
		h := sha256.Sum256([]byte(fmt.Sprintf("self-sdk-%s-%d", seed, time.Now().UnixNano())))
		return h[:]
	}
	return key
}

// setupClients demonstrates account initialization and configuration using the core SDK
// This function showcases how to:
// - Create Self SDK accounts with proper configuration
// - Set up storage paths for different account roles
// - Configure environment and logging settings
// - Handle account lifecycle management
func setupClients() (*account.Account, *account.Account) {
	fmt.Println("Setting up accounts...")

	issuerAccount, err := account.New(&account.Config{
		StorageKey:  generateStorageKey("issuer"),
		StoragePath: issuerStorageDir,
		Environment: account.TargetSandbox,
		LogLevel:    account.LogWarn,
	})
	if err != nil {
		log.Fatal("Failed to create issuer account:", err)
	}

	holderAccount, err := account.New(&account.Config{
		StorageKey:  generateStorageKey("holder"),
		StoragePath: holderStorageDir,
		Environment: account.TargetSandbox,
		LogLevel:    account.LogWarn,
	})
	if err != nil {
		log.Fatal("Failed to create holder account:", err)
	}

	return issuerAccount, holderAccount
}

// displayAccountInfo shows the unique DIDs for both parties
// DIDs (Decentralized Identifiers) are cryptographically verifiable identities
func displayAccountInfo(issuerAccount, holderAccount *account.Account) {
	issuerInbox, err := issuerAccount.InboxOpen()
	if err != nil {
		log.Fatal("Failed to open issuer inbox:", err)
	}
	holderInbox, err := holderAccount.InboxOpen()
	if err != nil {
		log.Fatal("Failed to open holder inbox:", err)
	}

	fmt.Printf("Issuer: %s\n", issuerInbox.String())
	fmt.Printf("   This is the credential issuer's unique decentralized identity\n")
	fmt.Printf("Holder: %s\n", holderInbox.String())
	fmt.Printf("   This is the credential holder's unique decentralized identity\n")
	fmt.Println()
}

// demonstrateBasicCredential showcases basic credential creation using the core SDK
// This example demonstrates:
// - Direct credential creation using credential.NewCredential()
// - Account-based credential issuance
// - Credential signing and issuance
// - Foundation concepts for all credential types
func demonstrateBasicCredential(issuerAccount, holderAccount *account.Account) {
	fmt.Println("1. Basic Email Credential")
	fmt.Println("=========================")

	issuerInbox, err := issuerAccount.InboxOpen()
	if err != nil {
		log.Fatal("Failed to open issuer inbox:", err)
	}
	holderInbox, err := holderAccount.InboxOpen()
	if err != nil {
		log.Fatal("Failed to open holder inbox:", err)
	}

	claims := map[string]interface{}{
		"emailAddress":     "john.doe@example.com",
		"verified":         true,
		"verificationDate": time.Now().Format("2006-01-02"),
	}

	credentialBuilder := credential.NewCredential().
		CredentialType(credential.CredentialTypeEmail).
		CredentialSubject(credential.AddressKey(holderInbox)).
		CredentialSubjectClaims(claims).
		Issuer(credential.AddressKey(issuerInbox)).
		ValidFrom(time.Now()).
		SignWith(issuerInbox, time.Now())

	unsignedCredential, err := credentialBuilder.Finish()
	if err != nil {
		log.Printf("Failed to build credential: %v", err)
		return
	}

	emailCredential, err := issuerAccount.CredentialIssue(unsignedCredential)
	if err != nil {
		log.Printf("Failed to issue credential: %v", err)
		return
	}

	fmt.Printf("✅ Email credential issued (Type: %v)\n", emailCredential.CredentialType())
	fmt.Printf("   Email: john.doe@example.com (verified)\n")
	fmt.Println()
}

// demonstrateProfileCredential showcases credentials with multiple claims using core SDK
// This example demonstrates:
// - Adding multiple claims to a single credential
// - Different data types in claims
// - Organizing related information in one credential
// - Building upon basic credential concepts
func demonstrateProfileCredential(issuerAccount, holderAccount *account.Account) {
	fmt.Println("2. Profile Credential")
	fmt.Println("=====================")

	issuerInbox, err := issuerAccount.InboxOpen()
	if err != nil {
		log.Fatal("Failed to open issuer inbox:", err)
	}
	holderInbox, err := holderAccount.InboxOpen()
	if err != nil {
		log.Fatal("Failed to open holder inbox:", err)
	}

	claims := map[string]interface{}{
		"firstName":        "John",
		"lastName":         "Doe",
		"displayName":      "John Doe",
		"profileLevel":     "verified",
		"country":          "United States",
		"registrationDate": time.Now().Format("2006-01-02"),
	}

	credentialBuilder := credential.NewCredential().
		CredentialType(credential.CredentialTypeProfileName).
		CredentialSubject(credential.AddressKey(holderInbox)).
		CredentialSubjectClaims(claims).
		Issuer(credential.AddressKey(issuerInbox)).
		ValidFrom(time.Now()).
		SignWith(issuerInbox, time.Now())

	unsignedCredential, err := credentialBuilder.Finish()
	if err != nil {
		log.Printf("Failed to build credential: %v", err)
		return
	}

	profileCredential, err := issuerAccount.CredentialIssue(unsignedCredential)
	if err != nil {
		log.Printf("Failed to issue credential: %v", err)
		return
	}

	fmt.Printf("✅ Profile credential issued (Type: %v)\n", profileCredential.CredentialType())
	fmt.Printf("   Name: John Doe, Country: United States\n")
	fmt.Println()
}

// demonstrateCustomCredentialWithEvidence showcases advanced credential features using core SDK
// This example demonstrates:
// - Creating custom credential types
// - Attaching file evidence to credentials using object.New()
// - Asset management and upload functionality
// - Creating verifiable presentations
// - Linking evidence to credential claims
func demonstrateCustomCredentialWithEvidence(issuerAccount, holderAccount *account.Account) {
	fmt.Println("3. Custom Credential with Evidence")
	fmt.Println("==================================")

	// Create evidence
	certificateData := []byte("Certificate of Completion\n" +
		"Advanced Go Programming Course\n" +
		"Student: John Doe\n" +
		"Grade: A+\n" +
		"Date: " + time.Now().Format("2006-01-02"))

	evidenceObj, err := object.New("application/pdf", certificateData)
	if err != nil {
		log.Printf("Failed to create evidence object: %v", err)
		return
	}

	err = issuerAccount.ObjectUpload(evidenceObj, false)
	if err != nil {
		log.Printf("Failed to upload evidence: %v", err)
		return
	}

	fmt.Printf("Evidence uploaded: %x (%d bytes)\n", evidenceObj.Id(), len(certificateData))

	issuerInbox, err := issuerAccount.InboxOpen()
	if err != nil {
		log.Fatal("Failed to open issuer inbox:", err)
	}
	holderInbox, err := holderAccount.InboxOpen()
	if err != nil {
		log.Fatal("Failed to open holder inbox:", err)
	}

	claims := map[string]interface{}{
		"certificationType": "Professional Development",
		"courseName":        "Advanced Go Programming",
		"completionDate":    time.Now().Format("2006-01-02"),
		"certificateHash":   fmt.Sprintf("%x", evidenceObj.Hash()),
		"grade":             "A+",
		"institution":       "Self SDK Academy",
		"courseHours":       40,
	}

	credentialBuilder := credential.NewCredential().
		CredentialType([]string{"VerifiableCredential", "CertificationCredential"}).
		CredentialSubject(credential.AddressKey(holderInbox)).
		CredentialSubjectClaims(claims).
		Issuer(credential.AddressKey(issuerInbox)).
		ValidFrom(time.Now()).
		SignWith(issuerInbox, time.Now())

	unsignedCredential, err := credentialBuilder.Finish()
	if err != nil {
		log.Printf("Failed to build credential: %v", err)
		return
	}

	customCredential, err := issuerAccount.CredentialIssue(unsignedCredential)
	if err != nil {
		log.Printf("Failed to issue credential: %v", err)
		return
	}

	presentation, err := createPresentation(issuerAccount, customCredential)
	if err != nil {
		log.Printf("Failed to create presentation: %v", err)
		return
	}

	fmt.Printf("✅ Certification credential issued (Type: %v)\n", customCredential.CredentialType())
	fmt.Printf("✅ Presentation created (Type: %v)\n", presentation.PresentationType())
	fmt.Printf("   Course: Advanced Go Programming, Grade: A+\n")
	fmt.Println()
}

// demonstrateOrganizationCredential showcases complex data structures in credentials using core SDK
// This example demonstrates:
// - Complex nested objects in claims
// - Arrays and collections in credentials
// - Hierarchical data organization
// - Real-world organizational data modeling
// - Advanced claim structuring techniques
func demonstrateOrganizationCredential(issuerAccount, holderAccount *account.Account) {
	fmt.Println("4. Organization Credential")
	fmt.Println("==========================")

	issuerInbox, err := issuerAccount.InboxOpen()
	if err != nil {
		log.Fatal("Failed to open issuer inbox:", err)
	}
	holderInbox, err := holderAccount.InboxOpen()
	if err != nil {
		log.Fatal("Failed to open holder inbox:", err)
	}

	claims := map[string]interface{}{
		"organizationName": "Acme Corporation",
		"employeeID":       "EMP-12345",
		"position":         "Senior Software Engineer",
		"department":       "Engineering",
		"startDate":        "2020-01-15",
		"address": map[string]interface{}{
			"street":  "123 Main Street",
			"city":    "San Francisco",
			"state":   "CA",
			"zipCode": "94105",
			"country": "United States",
		},
		"skills": []string{
			"Go Programming",
			"Microservices Architecture",
			"Docker & Kubernetes",
			"Cloud Computing",
			"API Design",
		},
		"certifications": []map[string]interface{}{
			{
				"name":       "AWS Solutions Architect",
				"issuer":     "Amazon Web Services",
				"issueDate":  "2021-03-15",
				"expiryDate": "2024-03-15",
				"level":      "Professional",
			},
			{
				"name":       "Kubernetes Administrator",
				"issuer":     "Cloud Native Computing Foundation",
				"issueDate":  "2021-08-20",
				"expiryDate": "2024-08-20",
				"level":      "Certified",
			},
		},
		"projects": []map[string]interface{}{
			{
				"name":         "Self SDK Integration",
				"role":         "Lead Developer",
				"duration":     "6 months",
				"technologies": []string{"Go", "Self SDK", "REST APIs"},
				"status":       "Completed",
			},
			{
				"name":         "Microservices Migration",
				"role":         "Senior Engineer",
				"duration":     "12 months",
				"technologies": []string{"Docker", "Kubernetes", "gRPC"},
				"status":       "In Progress",
			},
		},
		"performanceRating": map[string]interface{}{
			"overall":    "Exceeds Expectations",
			"technical":  "Expert",
			"leadership": "Strong",
			"reviewDate": time.Now().Format("2006-01-02"),
			"reviewer":   "Jane Smith, Engineering Manager",
		},
	}

	credentialBuilder := credential.NewCredential().
		CredentialType([]string{"VerifiableCredential", "EmploymentCredential"}).
		CredentialSubject(credential.AddressKey(holderInbox)).
		CredentialSubjectClaims(claims).
		Issuer(credential.AddressKey(issuerInbox)).
		ValidFrom(time.Now()).
		SignWith(issuerInbox, time.Now())

	unsignedCredential, err := credentialBuilder.Finish()
	if err != nil {
		log.Printf("Failed to build credential: %v", err)
		return
	}

	organizationCredential, err := issuerAccount.CredentialIssue(unsignedCredential)
	if err != nil {
		log.Printf("Failed to issue credential: %v", err)
		return
	}

	fmt.Printf("✅ Organization credential issued (Type: %v)\n", organizationCredential.CredentialType())
	fmt.Printf("   Employee: EMP-12345 - Senior Software Engineer\n")
	fmt.Printf("   Organization: Acme Corporation, Engineering Dept.\n")
	fmt.Println()
}

// createPresentation creates a verifiable presentation from a credential using core SDK
// This function demonstrates:
// - Presentation creation workflows
// - Credential packaging for sharing
// - Set presentation types and metadata
// - Prepare credentials for sharing with verifiers
func createPresentation(account *account.Account, cred *credential.VerifiableCredential) (*credential.VerifiablePresentation, error) {
	// 📋 PRESENTATION CREATION: Package credential for sharing using core SDK
	// Presentations allow selective disclosure of credential information

	// Get account inbox address for holder
	inboxAddress, err := account.InboxOpen()
	if err != nil {
		return nil, err
	}

	// Build the unsigned presentation
	builder := credential.NewPresentation().
		PresentationType([]string{"VerifiablePresentation"}).
		Holder(credential.AddressKey(inboxAddress)).
		CredentialAdd(cred)

	unsignedPresentation, err := builder.Finish()
	if err != nil {
		return nil, err
	}

	// Issue the presentation using the account
	return account.PresentationIssue(unsignedPresentation)
}

File: examples/server/02_credentials/01_issuing_credentials/05_comprehensive/java/Main.kt
/**
 * Demonstrates comprehensive credential issuance capabilities using the Self SDK.
 *
 * This example combines all previous concepts into a single demonstration:
 * - Basic credential issuance (e.g., email verification).
 * - Multi-claim credentials for richer data (e.g., user profiles).
 * - Credentials with linked file evidence (e.g., certifications).
 * - Credentials with complex, nested data structures (e.g., organizational roles).
 * - Creation of Verifiable Presentations to share credentials.
 *
 * 🎯 What you'll learn:
 * • A complete overview of the credential issuance lifecycle.
 * • Issuing various types of credentials, from simple to complex.
 * • Attaching file evidence to credentials using asset management.
 * • Structuring complex nested data and arrays within claims.
 * • Creating Verifiable Presentations to package and share credentials.
 */
import com.joinself.selfsdk.kmp.account.Account
import com.joinself.selfsdk.kmp.asset.BinaryObject
import com.joinself.selfsdk.kmp.credential.Address
import com.joinself.selfsdk.kmp.credential.CredentialBuilder
import com.joinself.selfsdk.kmp.credential.PresentationBuilder
import com.joinself.selfsdk.kmp.credential.VerifiableCredential
import com.joinself.selfsdk.kmp.credential.VerifiablePresentation
import com.joinself.selfsdk.kmp.time.Timestamp
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.concurrent.Semaphore

fun main() {
    println("Comprehensive Credential Issuance Demo")
    println("======================================")

    // Initialize issuer and holder accounts
    println("\nSetting up Self accounts...")
    val issuer = Common.setupAccount(storagePath = "./comprehensive_issuer_storage")
    val holder = Common.setupAccount(storagePath = "./comprehensive_holder_storage")
    println("✅ Accounts ready!")

    println()
    Common.displayAccountInfo(issuer, "Issuer")
    Common.displayAccountInfo(holder, "Holder")

    println("\nCREDENTIAL ISSUANCE EXAMPLES")
    println("============================")

    // Demonstrate each type of credential issuance
    demonstrateBasicCredential(issuer, holder)
    demonstrateProfileCredential(issuer, holder)
    demonstrateCustomCredentialWithEvidence(issuer, holder)
    demonstrateOrganizationCredential(issuer, holder)

    println("\n✅ All examples completed successfully!")
    println("\nPress enter to exit")
    readln()
}

/**
 * Demonstrates issuing a basic credential with a few simple claims.
 */
@OptIn(ExperimentalStdlibApi::class)
fun demonstrateBasicCredential(issuer: Account, holder: Account) {
    println("\n1. Basic Email Credential")
    println("=========================")

    val issuerAddress = Common.openInbox(issuer) ?: return
    val holderAddress = Common.openInbox(holder) ?: return

    val claims = mapOf(
        "emailAddress" to "john.doe@example.com",
        "verified" to true,
        "verificationDate" to LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE)
    )

    runCatching {
        val unsignedCredential = CredentialBuilder()
            .credentialType(arrayOf("EmailCredential"))
            .credentialSubject(Address.key(holderAddress))
            .credentialSubjectClaims(claims)
            .issuer(Address.key(issuerAddress))
            .validFrom(Timestamp.now())
            .signWith(issuerAddress, Timestamp.now())
            .finish()

        val emailCredential = issuer.credentialIssue(unsignedCredential)
        println("✅ Email credential issued (Type: ${emailCredential.credentialType().toList()})")
        println("   Email: john.doe@example.com (verified)")
    }.getOrElse { error ->
        println("❌ Failed to issue basic credential: ${error.message}")
    }
}

/**
 * Demonstrates issuing a credential with multiple top-level claims.
 */
@OptIn(ExperimentalStdlibApi::class)
fun demonstrateProfileCredential(issuer: Account, holder: Account) {
    println("\n2. Profile Credential")
    println("=====================")

    val issuerAddress = Common.openInbox(issuer) ?: return
    val holderAddress = Common.openInbox(holder) ?: return

    val claims = mapOf(
        "firstName" to "John",
        "lastName" to "Doe",
        "displayName" to "John Doe",
        "profileLevel" to "verified",
        "country" to "United States",
        "registrationDate" to LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE)
    )

    runCatching {
        val unsignedCredential = CredentialBuilder()
            .credentialType(arrayOf("ProfileNameCredential"))
            .credentialSubject(Address.key(holderAddress))
            .credentialSubjectClaims(claims)
            .issuer(Address.key(issuerAddress))
            .validFrom(Timestamp.now())
            .signWith(issuerAddress, Timestamp.now())
            .finish()

        val profileCredential = issuer.credentialIssue(unsignedCredential)
        println("✅ Profile credential issued (Type: ${profileCredential.credentialType().toList()})")
        println("   Name: John Doe, Country: United States")
    }.getOrElse { error ->
        println("❌ Failed to issue profile credential: ${error.message}")
    }
}

/**
 * Demonstrates issuing a credential with linked file evidence and creating a presentation.
 */
@OptIn(ExperimentalStdlibApi::class)
fun demonstrateCustomCredentialWithEvidence(issuer: Account, holder: Account) {
    println("\n3. Custom Credential with Evidence")
    println("==================================")

    val issuerAddress = Common.openInbox(issuer) ?: return
    val holderAddress = Common.openInbox(holder) ?: return

    // 1. Create and upload evidence asset
    val certificateData = """
    Certificate of Completion
    Advanced Kotlin Programming Course
    Student: John Doe
    Grade: A+
    Date: ${LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE)}
    """.trimIndent().toByteArray()

    val evidenceObj = runCatching {
        val signal = Semaphore(0)
        val obj = BinaryObject.create("application/pdf", certificateData)
        issuer.objectUpload(obj, true) { status ->
            signal.release()
        }
        signal.acquire()
        obj
    }.getOrElse { error ->
        println("❌ Failed to create or upload evidence: ${error.message}")
        return
    }
    println("✅ Evidence uploaded: ${evidenceObj.id()?.toHexString()} (${certificateData.size} bytes)")

    // 2. Create credential that links to the evidence
    val claims = mapOf(
        "certificationType" to "Professional Development",
        "courseName" to "Advanced Kotlin Programming",
        "completionDate" to LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE),
        "evidenceId" to (evidenceObj.id()?.toHexString() ?: ""),
        "grade" to "A+",
        "institution" to "Self SDK Academy",
        "courseHours" to 40
    )

    val customCredential = runCatching {
        val unsignedCredential = CredentialBuilder()
            .credentialType(arrayOf("VerifiableCredential", "CertificationCredential"))
            .credentialSubject(Address.key(holderAddress))
            .credentialSubjectClaims(claims)
            .issuer(Address.key(issuerAddress))
            .validFrom(Timestamp.now())
            .signWith(issuerAddress, Timestamp.now())
            .finish()
        issuer.credentialIssue(unsignedCredential)
    }.getOrElse { error ->
        println("❌ Failed to issue certification credential: ${error.message}")
        return
    }
    println("✅ Certification credential issued (Type: ${customCredential.credentialType().toList()})")

    // 3. Create a verifiable presentation containing the new credential
    val presentation = createPresentation(issuer, customCredential)
    if (presentation != null) {
        println("✅ Presentation created (Type: ${presentation.presentationType().toList()})")
        println("   Course: Advanced Kotlin Programming, Grade: A+")
    }
}

/**
 * Demonstrates issuing a credential with a complex, hierarchical data structure.
 */
@OptIn(ExperimentalStdlibApi::class)
fun demonstrateOrganizationCredential(issuer: Account, holder: Account) {
    println("\n4. Organization Credential (Complex Data)")
    println("=========================================")

    val issuerAddress = Common.openInbox(issuer) ?: return
    val holderAddress = Common.openInbox(holder) ?: return

    val claims = mapOf<String, Any>(
        "organizationName" to "Acme Corporation",
        "employeeID" to "EMP-12345",
        "position" to "Senior Software Engineer",
        "department" to "Engineering",
        "startDate" to "2020-01-15",
        "address" to mapOf(
            "street" to "123 Main Street",
            "city" to "San Francisco",
            "state" to "CA",
            "zipCode" to "94105",
            "country" to "United States"
        ),
        "skills" to listOf(
            "Kotlin Programming",
            "Microservices Architecture",
            "Docker & Kubernetes",
            "Cloud Computing",
            "API Design"
        ),
        "certifications" to listOf(
            mapOf(
                "name" to "AWS Solutions Architect",
                "issuer" to "Amazon Web Services",
                "issueDate" to "2021-03-15",
                "expiryDate" to "2024-03-15",
                "level" to "Professional"
            ),
            mapOf(
                "name" to "Kubernetes Administrator",
                "issuer" to "Cloud Native Computing Foundation",
                "issueDate" to "2021-08-20",
                "expiryDate" to "2024-08-20",
                "level" to "Certified"
            )
        ),
        "projects" to listOf(
            mapOf(
                "name" to "Self SDK Integration",
                "role" to "Lead Developer",
                "duration" to "6 months",
                "technologies" to listOf("Kotlin", "Self SDK", "REST APIs"),
                "status" to "Completed"
            ),
            mapOf(
                "name" to "Microservices Migration",
                "role" to "Senior Engineer",
                "duration" to "12 months",
                "technologies" to listOf("Docker", "Kubernetes", "gRPC"),
                "status" to "In Progress"
            )
        ),
        "performanceRating" to mapOf(
            "overall" to "Exceeds Expectations",
            "technical" to "Expert",
            "leadership" to "Strong",
            "reviewDate" to LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE),
            "reviewer" to "Jane Smith, Engineering Manager"
        )
    )

    runCatching {
        val unsignedCredential = CredentialBuilder()
            .credentialType(arrayOf("VerifiableCredential", "EmploymentCredential"))
            .credentialSubject(Address.key(holderAddress))
            .credentialSubjectClaims(claims)
            .issuer(Address.key(issuerAddress))
            .validFrom(Timestamp.now())
            .signWith(issuerAddress, Timestamp.now())
            .finish()

        val orgCredential = issuer.credentialIssue(unsignedCredential)
        println("✅ Organization credential issued (Type: ${orgCredential.credentialType().toList()})")
        println("   Employee: EMP-12345 - Senior Software Engineer")
        println("   Organization: Acme Corporation, Engineering Dept.")
    }.getOrElse { error ->
        println("❌ Failed to issue organization credential: ${error.message}")
    }
}

/**
 * Helper function to create a Verifiable Presentation from a credential.
 * @param account The account that will hold and sign the presentation.
 * @param cred The credential to include in the presentation.
 * @return The created VerifiablePresentation, or null on failure.
 */
fun createPresentation(account: Account, cred: VerifiableCredential): VerifiablePresentation? {
    println("   Creating verifiable presentation...")
    val holderAddress = Common.openInbox(account) ?: return null

    return runCatching {
        val unsignedPresentation = PresentationBuilder()
            .presentationType(arrayOf("VerifiablePresentation"))
            .holder(Address.key(holderAddress))
            .credentialAdd(cred)
            .finish()
        account.presentationIssue(unsignedPresentation)
    }.getOrElse { error ->
        println("   ❌ Failed to create presentation: ${error.message}")
        null
    }
}